src/storage.rs

```diff
@@ -126,136 +164,164 @@ impl MemStorageCore {
-        Ok(&self.snapshot)
+        snapshot
     }
 
     /// Discards all log entries prior to compact_index.
     /// It is the application's responsibility to not attempt to compact an index
     /// greater than RaftLog.applied.
+    ///
+    /// # Panics
+    ///
+    /// Panics if `compact_index` is higher than `Storage::last_index(&self) + 1`.
     pub fn compact(&mut self, compact_index: u64) -> Result<()> {
-        let offset = self.entries[0].get_index();
-        if compact_index <= offset {
-            return Err(Error::Store(StorageError::Compacted));
+        if compact_index <= self.inner_first_index() {
+            // Don't need to treat this case as an error.
+            return Ok(());
         }
-        if compact_index > self.inner_last_index() {
+
+        if compact_index > self.inner_last_index() + 1 {
             panic!(
-                "compact {} is out of bound lastindex({})",
+                "compact not received raft logs: {}, last index: {}",
                 compact_index,
                 self.inner_last_index()
-            )
+            );
         }
 
-        let i = (compact_index - offset) as usize;
-        let entries = self.entries.drain(i..).collect();
-        self.entries = entries;
+        if let Some(entry) = self.entries.first() {
+            let offset = compact_index - entry.get_index();
+            self.entries.drain(..offset as usize);
+        }
         Ok(())
     }
 
     /// Append the new entries to storage.
-    /// TODO: ensure the entries are continuous and
-    /// entries[0].get_index() > self.entries[0].get_index()
     pub fn append(&mut self, ents: &[Entry]) -> Result<()> {
         if ents.is_empty() {
             return Ok(());
         }
-        let first = self.entries[0].get_index() + 1;
-        let last = ents[0].get_index() + ents.len() as u64 - 1;
-
-        if last < first {
-            return Ok(());
-        }
-        // truncate compacted entries
-        let te: &[Entry] = if first > ents[0].get_index() {
-            let start_ent = (first - ents[0].get_index()) as usize;
-            &ents[start_ent..]
-        } else {
-            ents
-        };
-
-        let offset = te[0].get_index() - self.entries[0].get_index();
-        if self.entries.len() as u64 > offset {
-            let mut new_entries: Vec<Entry> = vec![];
-            new_entries.extend_from_slice(&self.entries[..offset as usize]);
-            new_entries.extend_from_slice(te);
-            self.entries = new_entries;
-        } else if self.entries.len() as u64 == offset {
-            self.entries.extend_from_slice(te);
-        } else {
+        if self.inner_first_index() > ents[0].get_index() {
             panic!(
-                "missing log entry [last: {}, append at: {}]",
-                self.inner_last_index(),
-                te[0].get_index()
-            )
+                "overwrite compacted raft logs, compacted: {}, append: {}",
```

Comment:
What if messages are reordered?

Reply:
You mean `ents` is not ordered? A correct Raft ensures it's ordered.

Reply:
If you mean some raft messages are delayed in network, It's Raft's responsibility to truncate redundant entries.

Reply:
Fixed.